skimage/restoration/_cycle_spin.py

```diff
@@ -0,0 +1,157 @@
+
+import numpy as np
+
+
+def _generate_shifts(ndim, max_shifts=None, shift_steps=1):
+    """Returns all combinations of shifts in n dimensions over the specified
+    max_shifts and step sizes.
+
+    Examples
+    --------
+    print(list(_generate_shifts(2, max_shifts=(1, 2), shift_steps=1)))
+    >>> [(0, 0), (0, 1), (0, 2), (1, 0), (1, 1), (1, 2)]
+    """
+    if max_shifts is None:
+        max_shifts = (0, )*ndim
+    elif np.isscalar(max_shifts):
+        max_shifts = (max_shifts, )*ndim
+    elif len(max_shifts) != ndim:
+        raise ValueError("max_shifts should have length ndim")
+
+    if shift_steps is None:
+        shift_steps = (1, )*ndim
+    elif np.isscalar(shift_steps):
+        shift_steps = (shift_steps, )*ndim
+    elif len(shift_steps) != ndim:
+        raise ValueError("max_shifts should have length ndim")
+
+    if np.any(np.asarray(shift_steps) < 1):
+        raise ValueError("shift_steps must all be >= 1")
+    return product(*([range(0, s+1, t) for
+                      s, t in zip(max_shifts, shift_steps)]))
+
+
+def _roll_axes(x, rolls, axes=None):
+    """Apply np.roll along a set of axes.
+
+    Parameters
+    ----------
+    x : array-like
+        array to roll
+    rolls : int or sequence
+        The amount to roll along each axis in ``axes``
+    axes : int or sequence
+        The axes to roll.  Default is the first len(roll) axes of x.
+
+    Returns
+    -------
+    x : array
+        data with axes rolled
+    """
+    if np.isscalar(rolls):
+        rolls = (rolls, )
+    if len(rolls) > x.ndim:
+        raise ValueError("length of rolls should not exceed x.ndim")
+    if axes is None:
+        axes = np.arange(len(rolls))
+    if np.isscalar(axes):
+        axes = (axes, )
+    if len(rolls) != len(axes):
+        raise ValueError("size mismatch between rolls and axes")
+    try:
+        # numpy>=1.12 supports tuple for rolls, axes
+        return np.roll(x, rolls, axes)
+    except TypeError:
+        for r, a in zip(rolls, axes):
+            x = np.roll(x, r, a)
+    return x
+
+
+def cycle_spin(x, func, max_shifts, shift_steps=1, num_workers=1,
+               func_kw={}):
+    """Cycle spinning (repeatedly apply func to shifted versions of x).
+
+    Parameters
+    ----------
+    x : array-like
+        data
+    func : function
+        A function to apply to circularly shifted versions of x.  Should take
+        x as its first argument.  Any additional arguments can be supplied via
+        ``func_kw``.
+    max_shifts : int or tuple, optional
+        If an integer, shifts in ``range(0, max_shifts+1)`` will be used along
+        each axis of x. If a tuple, ``range(0, max_shifts[i]+1)`` will be used
+        along axis i.
+    shift_steps : int or tuple, optional
+        The step size for the shifts applied along axis, i, are::
+        ``range((0, max_shifts[i]+1, shift_steps[i]))``.  If an integer is
+        provided, the same step size is used for all axes.
```

Comment:
num_workers is not documented. How about a None value would put the number of CPUs? We use dask as an optional dependency. What would be the advantage of concurrency instead of dask?